### PR TITLE
Add controlplane controller for Openstack

### LINKED ISSUE #56: Implement `ControlPlane` controller for OpenStack provider
Similar to how we have implemented the `ControlPlane` extension resource controller for the [AWS provider](https://github.com/gardener/gardener-extensions/tree/master/controllers/provider-aws/pkg/controller/controlplane) let's please now do it for OpenStack.

Based on [the current implementation](https://github.com/gardener/gardener/blob/master/example/90-shoot-openstack.yaml) the `ControlPlaneConfig` should look like this:

```yaml
apiVersion: openstack.provider.extensions.gardener.cloud/v1alpha1
kind: ControlPlaneConfig
cloudControllerManager:
  featureGates:
    CustomResourceValidation: true
loadBalancerProvider: haproxy
```

No `ControlPlaneStatus` needs to be implemented right now (not needed yet).

### PR DESCRIPTION
**Pull request by svetlinas**
**What this PR does / why we need it:**
Adds controlplane webhooks for Alicloud.

**Which issue(s) this PR fixes:**
Fixes #56 

**Special notes for your reviewer:**
The code follows the same logic as the other providers that has it implemented already.

**Release note:**
Add Openstack controlplane controller

### CODE REVIEW

File: controllers/provider-openstack/charts/internal/cloud-controller-manager/templates/cloud-controller-manager.yaml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            - /hyperkube
+            - cloud-controller-manager
+            - --allocate-node-cidrs=true
+            - --cloud-provider=openstack
+            - --cloud-config=/etc/kubernetes/cloudprovider/cloudprovider.conf
+            - --cluster-cidr={{ .Values.podNetwork }}
+            - --cluster-name={{ .Values.clusterName }}
+            - --concurrent-service-syncs=10
+            - --configure-cloud-routes=true
+                  {{- include "cloud-controller-manager.featureGates" . | trimSuffix "," | indent 8 }}
```
Comment (stoyanr) [inline controllers/provider-openstack/charts/internal/cloud-controller-manager/templates/cloud-controller-manager.yaml]: These should be aligned from the beginning of the line, as in other templates.

File: controllers/provider-openstack/charts/internal/cloud-provider-config/values.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+kubernetesVersion: version
+#[Global]
+authUrl: https://keystone/v3/
+domainName: %q
```
Comment (stoyanr) [inline controllers/provider-openstack/charts/internal/cloud-provider-config/values.yaml]: These should contain concrete values, either reasonable defaults or example values.

File: controllers/provider-openstack/charts/internal/cloud-provider-config/templates/cloud-provider-config.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    [LoadBalancer]
+    lb-version=v2
+    lb-provider: {{ .Values.lbProvider }}
+    floating-network-id: {{ .Values.floatingNetworkID }}
+    subnet-id: {{ .Values.subnetID }}
+    create-monitor=true
+    monitor-delay=60s
+    monitor-timeout=30s
+    monitor-max-retries=5
+    {{- if and (semverCompare ">= 1.10.1" .Values.kubernetesVersion) (semverCompare "< 1.10.3" .Values.kubernetesVersion) }}
```
Comment (stoyanr) [inline controllers/provider-openstack/charts/internal/cloud-provider-config/templates/cloud-provider-config.yaml]: Not sure if we should support Kubernetes 1.10.x in extension controllers, let's ask.

File: controllers/provider-openstack/pkg/apis/openstack/types_controlplane.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+	// LoadBalancerProvider is the name of the load balancer provider in the OpenStack environment.
+	LoadBalancerProvider string
+
+	// DHCPDomain is the dhcp domain of the OpenStack system configured in nova.conf. Only meaningful for
+	// Kubernetes 1.10.1+. See https://github.com/kubernetes/kubernetes/pull/61890 for details.
+	// +optional
+	DHCPDomain *string
+	// RequestTimeout specifies the HTTP timeout against the OpenStack API.
+	// +optional
```
Comment (stoyanr) [inline controllers/provider-openstack/pkg/apis/openstack/types_controlplane.go]: The tag `+optional` is only needed for concrete versions, e.g. `v1alpha1`.

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+		"password":          c.Password,
+		"lbProvider":        cpConfig.LoadBalancerProvider,
+		"floatingNetworkID": infraStatus.Networks.FloatingPool.ID,
+		"subnetID":          getSubnetID(infraStatus),
+		"dhcpDomain":        cpConfig.DHCPDomain,
+		"requestTimeout":    cpConfig.RequestTimeout,
+	}, nil
+}
+
+// TODO should we return error in case the subnetID is nil?
```
Comment (stoyanr) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: Yes, we should. The logic for AWS and Alicloud is similar, see how it's implemented there.

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider_test.go [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+			},
+			Spec: extensionsv1alpha1.ControlPlaneSpec{
+				SecretRef: corev1.SecretReference{
+					Name:      common.CloudProviderSecretName,
+					Namespace: namespace,
+				},
+				ProviderConfig: &runtime.RawExtension{
+					Raw: encode(&openstackv1aplha.ControlPlaneConfig{
+						LoadBalancerProvider: "load-balancer-provider",
+						DHCPDomain:           &dhcpDomain,
```
Comment (stoyanr) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider_test.go]: You could use `util.StringPtr` to convert a string literal, e.g. "dhcp-domain" into a pointer directly.

File: controllers/provider-openstack/pkg/apis/openstack/types_controlplane.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	RequestTimeout *string
```
Comment (stoyanr) [inline controllers/provider-openstack/pkg/apis/openstack/types_controlplane.go]: The type of a timeout should not be `string`, but rather `metav1.Duration`. Kubernetes will do correct marshalling / unmarshalling to and from YAML for this type.

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider_test.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+						RequestTimeout:       &requestTimeout,
```
Comment (stoyanr) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider_test.go]: As mentioned above, the type of `RequestTimeout` should be `metav1.Duration`. To easily get a pointer from a value, you could add a function `util.DurationPtr`.

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+			Type: &corev1.ConfigMap{},
+			Name: openstack.CloudProviderConfigName,
+		},
+	},
+}
+
+var ccmChart = &chart.Chart{
+	Name: "cloud-controller-manager",
+	Path: filepath.Join(openstack.InternalChartsPath, "cloud-controller-manager"),
+	// TODO Images: []string{internal.HyperkubeImageName},
```
Comment (stoyanr) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: You can add the `hyperkube` image to `images.yaml`, see how this is done for other providers.

File: controllers/provider-openstack/charts/internal/cloud-provider-config/templates/cloud-provider-config.yaml (hunk 3 above)
Comment (svetlinas) [reply controllers/provider-openstack/charts/internal/cloud-provider-config/templates/cloud-provider-config.yaml]: Based on the discussion in Slack, version 1.10 will continue to be supported.

File: controllers/provider-openstack/example/30-controlplane.yaml [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    kind: CloudProfile
+  seed:
+    apiVersion: garden.sapcloud.io/v1beta1
+    kind: Seed
+  shoot:
+    apiVersion: garden.sapcloud.io/v1beta1
+    kind: Shoot
+    spec:
+      cloud:
+        openstack:
```
Comment (rfranzke) [inline controllers/provider-openstack/example/30-controlplane.yaml]: Shouldn't this example contain the keyStoneURL?

File: controllers/provider-openstack/pkg/apis/openstack/types_controlplane.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+// ControlPlaneConfig contains configuration settings for the control plane.
+type ControlPlaneConfig struct {
+	metav1.TypeMeta
+
+	// LoadBalancerProvider is the name of the load balancer provider in the OpenStack environment.
+	LoadBalancerProvider string
+
+	// KeyStoneURL is the URL for auth{n,z} in OpenStack (pointing to KeyStone).
+	KeyStoneURL string
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/apis/openstack/types_controlplane.go]: This should be read out of the `CloudProfile`

File: controllers/provider-openstack/example/30-infrastructure.yaml [hunk 12]
```diff
@@ -62,4 +62,3 @@ spec:
       worker: '10.250.0.0/19'
     zones:
     - name: zone_1_1
-  sshPublicKey: ...
```
Comment (rfranzke) [inline controllers/provider-openstack/example/30-infrastructure.yaml]: Why is this removed?

File: controllers/provider-openstack/pkg/apis/openstack/v1alpha1/types_controlplane.go [hunk 13]
```diff
@@ -0,0 +1,35 @@
+package v1alpha1
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/apis/openstack/v1alpha1/types_controlplane.go]: license header is missing

File: controllers/provider-openstack/pkg/apis/openstack/v1alpha1/types_controlplane.go [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+// ControlPlaneConfig contains configuration settings for the control plane.
+type ControlPlaneConfig struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// LoadBalancerProvider is the name of the load balancer provider in the OpenStack environment.
+	LoadBalancerProvider string `json:"loadBalancerProvider"`
+
+	// KeyStoneURL is the URL for auth{n,z} in OpenStack (pointing to KeyStone).
+	KeyStoneURL string `json:"keystoneURL"`
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/apis/openstack/v1alpha1/types_controlplane.go]: See above, should be read out of the cloud profile

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go [hunk 15]
```diff
@@ -0,0 +1,250 @@
+package controlplane
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: license header

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+	if cpConfig.CloudControllerManager != nil {
+		values["featureGates"] = cpConfig.CloudControllerManager.FeatureGates
+	}
+
+	return values, nil
+}
+
+// getSubnetID determines the subnet ID from the given infrastructure status by looking for the first
+// subnet with purpose "nodes".
+func getSubnetID(infraStatus *openstackv1alpha1.InfrastructureStatus) (string, error) {
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: Use https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-openstack/pkg/apis/openstack/helper/helper.go#L26

File: controllers/provider-openstack/pkg/apis/openstack/types_controlplane.go [hunk 17]
```diff
@@ -0,0 +1,31 @@
+package openstack
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/apis/openstack/types_controlplane.go]: license header is missing

**Comment by rfranzke** [comment]
Also `make generate` needs to be run.
